src/layoutrestorer/layoutrestorer.ts

```diff
@@ -0,0 +1,249 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) Jupyter Development Team.
+| Distributed under the terms of the Modified BSD License.
+|----------------------------------------------------------------------------*/
+
+import {
+  JSONObject
+} from 'phosphor/lib/algorithm/json';
+
+import {
+  AttachedProperty
+} from 'phosphor/lib/core/properties';
+
+import {
+  defineSignal, ISignal
+} from 'phosphor/lib/core/signaling';
+
+import {
+  Widget
+} from 'phosphor/lib/ui/widget';
+
+import {
+  Token
+} from 'phosphor/lib/core/token';
+
+import {
+  IStateDB
+} from '../statedb';
+
+
+/* tslint:disable */
+/**
+ * The layout restorer token.
+ */
+export
+const ILayoutRestorer = new Token<ILayoutRestorer>('jupyter.services.layout-restorer');
+/* tslint:enable */
+
+
+/**
+ * A static class that restores the layout of the application when it reloads.
+ */
+export
+interface ILayoutRestorer {
+  /**
+   * Add a widget to be tracked by the layout restorer.
+   */
+  add(widget: Widget, name: string): void;
+
+  /**
+   * Wait for the given promise to resolve before restoring layout.
+   *
+   * #### Notes
+   * This function should only be called before the `first` promise passed in
+   * at instantiation has resolved. See the notes for `LayoutRestorer.IOptions`.
+   */
+  await(promise: Promise<any>): void;
```

Comment:
It's just like `delete`, using a key word as a method name when it makes semantic sense seems like a good idea to me.